src/libmongoc/src/mongoc/mongoc-async-cmd.c

```diff
@@ -134,20 +136,38 @@ mongoc_async_cmd_run (mongoc_async_cmd_t *acmd)
 }
 
 void
-_mongoc_async_cmd_init_send (mongoc_async_cmd_t *acmd, const char *dbname)
+_mongoc_async_cmd_init_send (const mongoc_opcode_t cmd_opcode, mongoc_async_cmd_t *acmd, const char *dbname)
 {
+   acmd->rpc.header.msg_len = 0; 
+   acmd->rpc.header.request_id = ++acmd->async->request_id; /* in _server_monitor_awaitable_hello_send(), this is postfix++ */
+   acmd->rpc.header.response_to = 0;
+
+   if(MONGOC_OPCODE_QUERY == cmd_opcode) {
+
    acmd->ns = bson_strdup_printf ("%s.$cmd", dbname);
 
-   acmd->rpc.header.msg_len = 0;
-   acmd->rpc.header.request_id = ++acmd->async->request_id;
-   acmd->rpc.header.response_to = 0;
-   acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;
+   acmd->rpc.header.opcode = MONGOC_OPCODE_QUERY;    
    acmd->rpc.query.flags = MONGOC_QUERY_SECONDARY_OK;
    acmd->rpc.query.collection = acmd->ns;
    acmd->rpc.query.skip = 0;
    acmd->rpc.query.n_return = -1;
    acmd->rpc.query.query = bson_get_data (&acmd->cmd);
    acmd->rpc.query.fields = NULL;
+   }
+
+
+if(MONGOC_OPCODE_MSG == cmd_opcode) { 
+
+   acmd->rpc.header.opcode = MONGOC_OPCODE_MSG;    
+
+   acmd->rpc.msg.msg_len = 0;
+   acmd->rpc.msg.flags = 0; /* JFW: other flags, like MONGOC_MSG_EXHAUST_ALLOWED, as per _server_monitor_awaitable_hello_send(), fail validation */
```

Comment:
`MONGOC_MSG_EXHAUST_ALLOWED` does not apply for single-threaded monitoring. 

`MONGOC_MSG_EXHAUST_ALLOWED` corresponds to `exhaustAllowed` in [OP_MSG flags](https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/#op_msg)

> The client is prepared for multiple replies to this request using the moreToCome bit. The server will never produce replies with the moreToCome bit set unless the request has this bit set.
>
> This ensures that multiple replies are only sent when the network layer of the requester is prepared for them.

For single-threaded monitoring, the requester is not prepared for multiple replies. After `_mongoc_async_cmd_phase_recv_rpc` completes reading one `OP_REPLY` or `OP_MSG`, it does not continue reading replies. Nor should it. Single-threaded monitoring blocks the main thread to send a `hello` / legacy hello to all servers. 

By "fail validation", I assume the server returned an error when this flag was set. That is likely due to the server expecting `topologyVersion` to be included in the [awaitable hello command](https://github.com/mongodb/specifications/blob/6dc6f80026f0f8d99a8c81f996389534b14f6602/source/server-discovery-and-monitoring/server-monitoring.rst#awaitable-hello-or-legacy-hello-server-specification).